commit a2dfa06b5720c2a40ddcc00d4e7fa528d9072f1f
Author: Tero Marttila <terom@fixme.fi>
Date:   Sat Feb 5 18:06:05 2022 +0200

    tools: fix idf_monitor newline handling

diff --git a/tools/idf_monitor_base/serial_handler.py b/tools/idf_monitor_base/serial_handler.py
index af267088f2..5a3a9d98df 100644
--- a/tools/idf_monitor_base/serial_handler.py
+++ b/tools/idf_monitor_base/serial_handler.py
@@ -78,12 +78,9 @@ class SerialHandler:
             # add unprocessed part from previous "data" to the first line
             sp[0] = self._last_line_part + sp[0]
             self._last_line_part = b''
-        if sp[-1] != b'':
-            # last part is not a full line
-            self._last_line_part = sp.pop()
+        # last part is not a full line
+        self._last_line_part = sp.pop()
         for line in sp:
-            if line == b'':
-                continue
             if self._serial_check_exit and line == console_parser.exit_key.encode('latin-1'):
                 raise SerialStopException()
             if gdb_helper:
